### DISCUSSION TITLE
Environment variables and deploying Nuxt 3 on a local server using PM2

### DISCUSSION THREAD

**Discussion by xdiegom** [nuxt/nuxt#26828] [Questions] [answered]
Hi everyone, hope you are doing well! 👋🏼

I am currently working on a nuxt 3 project in which the contractual statements with the client say that the code cannot be shared and only the compilation files can be shared so they can be installed in their own servers.

So, because they don't use cloud server provider, one of the proposed solutions is to use pm2 in order to run the project with only the generated .output. files and using the `ecosystem.config.cjs` file that [nuxt documentation suggests](https://nuxt.com/docs/getting-started/deployment#pm2).

My question is: 

> Will/Could PM2 override the environment variables that have been embedded into the ./output directory as a result of the project being built with `npx nuxi build` or `npm run build` with the environment variables defined in `ecosystem.config.cjs` file? E.g:

```js
module.exports = {
  apps: [
    {
      name: "pm2-app",
      port: "3000",
      exec_mode: "cluster",
      instances: "max",
      script: "./.output/server/index.mjs",
      env: {
        NUXT_FOO_KEY: "DEV_YOLO",
      },
      env_production: {
        NUXT_FOO_KEY: "PROD_YOLO",
      },
    },
  ],
};
```

### I want to know if this is possible or not, if its something that I ignore about using PM2. Why?:

Because, what I am facing is that the client doesn't want to share the production keys so I can build the app and deliver the compiled project. 

I have a reproduction [demo repository](https://github.com/xdiegom/pm2-nuxt) for this scenario that you can clone, so after run pm2 with one of the options described in the `README` file, I expect to see either `DEV_YOLO` or `PROD_YOLO` when you visit `http://localhost:3000`.

I am eager to know what are your thoughts or ideas about this situation and if there is a workaround for it! 😃

**Comment by xdiegom** [comment]
Forget it, I just solved the problem.

`NUXT_FOO_KEY` should be with the prefix `NUXT_PUBLIC_`, something that I missed from the documentation that I didn't pay attention. 

Then, in `nuxt.config.ts`, the runtimeConfig attribute with public keys, the variable is referenced as [documentation](https://nuxt.com/docs/guide/going-further/runtime-config#environment-variables) recommends:

```
// .env
NUXT_PUBLIC_FOO_KEY=
```
Then: 
```
// nuxt.config.ts
runtimeConfig: {
   fooKey: ''
}
```

PM2 actually uses the environment variables and pass them to the node server!

**Reply by phuze** [reply]
`NUXT_PUBLIC_` as it suggests, will expose those variables to the client. You should absolutely **not** prefix any sensitive or private keys with `NUXT_PUBLIC_`.

**Comment by dsvgl** [comment]
Must be `NUXT_PUBLIC_FOO_KEY` instead of `NUXT_FOO_KEY` in your ecosystem file. Then it will work 😊

Note, that you **must** prepend your vars with `NUXT` to have runtimeConfig vars replaced with your env vars at runtime.
Just using `FOO_KEY` for example will not work.

> Only a specially-named environment variable can override a runtime config property. That is, an uppercase environment variable starting with **NUXT_** which uses **_** to separate keys and case changes.

see: https://nuxt.com/docs/guide/going-further/runtime-config#environment-variables

**Reply by dsvgl** [reply, marked as answer]
ah, you already got it yourself 😄

**Reply by xdiegom** [reply]
> ah, you already got it yourself 😄 

Exactly! 

Thanks man for your help!

**Comment by kaliloulah1155** [comment]
1- Install dotenv
npm i dotenv
2-Make configuration in your file nuxt.config.ts
   require('dotenv').config();
    runtimeConfig: {
        public:{
            tva:process.env.NUXT_PUBLIC_TVA
        }
    },
   3- your file .env must be 
     NUXT_PUBLIC_TVA=10
     
     let nuxt control your env instead of pm2
